test/varmat_compatibility.py

```diff
@@ -0,0 +1,260 @@
+    work_queue = Queue.Queue()
+
+    # For each signature, generate cpp code to test
+    for n, signature in enumerate(signatures_to_check):
+        sp = SignatureParser(signature)
+
+        if sp.is_high_order():
+            work_queue.put((n, signature, None))
+            continue
+
+        cpp_code = ""
+        any_overload_uses_varmat = False
+        
+        for m, overloads in enumerate(itertools.product(("Prim", "Rev", "RevVarmat"), repeat = sp.number_arguments())):
+            cg = CodeGenerator()
+
+            arg_list_base = cg.build_arguments(sp, overloads, size = 1)
+
+            arg_list = []
+            for overload, arg in zip(overloads, arg_list_base):
+                if arg.is_reverse_mode() and arg.is_varmat_compatible() and overload.endswith("Varmat"):
+                    any_overload_uses_varmat = True
+                    arg = cg.to_var_value(arg)
+                
+                arg_list.append(arg)
+
+            cg.function_call_assign("stan::math::" + sp.function_name, *arg_list)
+
+            cpp_code += BENCHMARK_TEMPLATE.format(
+                benchmark_name = sp.function_name + repr(m),
+                code=cg.cpp(),
+            )
+
+        if any_overload_uses_varmat:
+            work_queue.put((n, signature, cpp_code))
+        else:
+            work_queue.put((n, signature, None))
+
+    output_lock = threading.Lock()
+
+    if not os.path.exists(WORKING_FOLDER):
+        os.mkdir(WORKING_FOLDER)
+
+    # Test if each cpp file builds and update the output file
+    # This part is done in parallel
+    def worker():
+        while True:
+            try:
+                n, signature, cpp_code = work_queue.get(False)
+            except Queue.Empty:
+                return # If queue is empty, worker quits
+
+            # Test the signature
+            successful, attempted = build_signature(cpp_code)
+
+            # Acquire a lock to do I/O
+            with output_lock:
+                if attempted:
+                    if successful:
+                        result_string = "Success!"
+                        compatible_signatures.add(signature)
+                    else:
+                        result_string = "Fail!"
+                        incompatible_signatures.add(signature)
+                else:
+                    result_string = "Impossible!"
+                    impossible_signatures.add(signature)
+
+                print("Results of test {0} / {1}, {2} ... ".format(n, len(signatures_to_check), signature.strip()) + result_string)
+
+                if results_file:
+                    with open(results_file, "w") as f:
+                        json.dump({
+                            "compatible_signatures" : list(compatible_signatures),
+                            "incompatible_signatures" : list(incompatible_signatures),
+                            "impossible_signatures" : list(impossible_signatures)
+                        }, f, indent = 4, sort_keys = True)
```

Comment:
Is every worker rewriting whole json every time it completes work on one signature? That is inefficient.

Reply:
Well, it guarantees the json file is always updated and it's only happening every few seconds even with lots of parallelism (for each signature we're compiling a file). So inefficient, but I prefer the partial output to waiting till the end to print.

Reply:
Hmm, I guess I'm not really using the partial output. Maybe this was more important when this script regularly crashed in development. Now it's not supposed to regularly crash at least lol. I moved the write out.